File: src/sync/merge.py
"""
Three-way merge for text files using diff3 algorithm.
"""

import bisect
import difflib
from dataclasses import dataclass
from typing import List, Optional, Tuple


@dataclass
class MergeResult:
    merged_text: str
    has_conflicts: bool
    conflict_count: int


def _blocks_to_edits(
    base_len: int,
    other_len: int,
    matching_blocks: List[Tuple[int, int, int]],
) -> List[Tuple[int, int, int, int, bool]]:
    edits = []
    prev_i, prev_j = 0, 0
    for i, j, n in matching_blocks:
        if prev_i < i or prev_j < j:
            edits.append((prev_i, i, prev_j, j, False))
        if n > 0:
            edits.append((i, i + n, j, j + n, True))
        prev_i, prev_j = i + n, j + n
    if prev_i < base_len or prev_j < other_len:
        edits.append((prev_i, base_len, prev_j, other_len, False))
    return edits


class _EditIndex:
    """Binary-searchable index over sorted edit spans for O(log E) lookups."""

    __slots__ = ("_edits", "_starts")

    def __init__(self, edits: List[Tuple[int, int, int, int, bool]]):
        self._edits = edits
        self._starts = [e[0] for e in edits]

    def find(self, lo: int, hi: int) -> Optional[Tuple[int, int, bool]]:
        if lo == hi:
            return self._find_insertion_point(lo)
        return self._find_range(lo, hi)

    def _find_insertion_point(self, lo: int) -> Optional[Tuple[int, int, bool]]:
        idx = bisect.bisect_right(self._starts, lo) - 1
        search_range = range(max(0, idx - 1), min(len(self._edits), idx + 3))
        for i in search_range:
            base_lo, base_hi, other_lo, other_hi, is_same = self._edits[i]
            if base_lo == lo and base_hi == lo and not is_same:
                return (other_lo, other_hi, False)
        for i in search_range:
            base_lo, base_hi, other_lo, other_hi, is_same = self._edits[i]
            if base_lo <= lo <= base_hi and is_same:
                mapped = other_lo + (lo - base_lo)
                return (mapped, mapped, True)
        return None

    def _find_range(self, lo: int, hi: int) -> Optional[Tuple[int, int, bool]]:
        idx = bisect.bisect_right(self._starts, lo) - 1
        for i in range(max(0, idx - 1), min(len(self._edits), idx + 3)):
            base_lo, base_hi, other_lo, other_hi, is_same = self._edits[i]
            if base_lo <= lo and hi <= base_hi:
                if is_same:
                    return (other_lo + (lo - base_lo), other_lo + (hi - base_lo), True)
                return (other_lo, other_hi, False)
        return None


def three_way_merge(base: str, ours: str, theirs: str) -> MergeResult:
    if base is None or ours is None or theirs is None:
        raise TypeError("base, ours, and theirs must be strings, not None")
    base_lines = base.splitlines(keepends=True) if base else []
    ours_lines = ours.splitlines(keepends=True) if ours else []
    theirs_lines = theirs.splitlines(keepends=True) if theirs else []

    sm_ours = difflib.SequenceMatcher(None, base_lines, ours_lines)
    sm_theirs = difflib.SequenceMatcher(None, base_lines, theirs_lines)
    edits_ours = _blocks_to_edits(len(base_lines), len(ours_lines), sm_ours.get_matching_blocks())
    edits_theirs = _blocks_to_edits(
        len(base_lines), len(theirs_lines), sm_theirs.get_matching_blocks()
    )
    idx_ours = _EditIndex(edits_ours)
    idx_theirs = _EditIndex(edits_theirs)

    break_points = {0, len(base_lines)}
    for e in edits_ours:
        break_points.add(e[0])
        break_points.add(e[1])
    for e in edits_theirs:
        break_points.add(e[0])
        break_points.add(e[1])
    pts = sorted(break_points)

    segments: List[Tuple[int, int]] = []
    for i in range(len(pts)):
        if i + 1 < len(pts):
            segments.append((pts[i], pts[i]))      # zero-length (insertion point)
            segments.append((pts[i], pts[i + 1]))   # normal range
        else:
            segments.append((pts[i], pts[i]))        # trailing insertion point
    if not segments:
        segments = [(0, 0)]

    output: List[str] = []
    conflict_count = 0

    for lo, hi in segments:
        if lo > hi:
            continue
        ours_info = idx_ours.find(lo, hi)
        theirs_info = idx_theirs.find(lo, hi)
        if ours_info is None or theirs_info is None:
            output.extend(base_lines[lo:hi])
            continue

        ours_lo, ours_hi, ours_same = ours_info
        theirs_lo, theirs_hi, theirs_same = theirs_info
        ours_content = ours_lines[ours_lo:ours_hi]
        theirs_content = theirs_lines[theirs_lo:theirs_hi]

        if ours_same and theirs_same:
            output.extend(base_lines[lo:hi])
        elif ours_same and not theirs_same:
            output.extend(theirs_content)
        elif not ours_same and theirs_same:
            output.extend(ours_content)
        else:
            if ours_content == theirs_content:
                output.extend(ours_content)
            else:
                conflict_count += 1
                output.append("<<<<<<< LOCAL\n")
                output.extend(ours_content)
                output.append("=======\n")
                output.extend(theirs_content)
                output.append(">>>>>>> CLOUD\n")

    merged = "".join(output)
    return MergeResult(
        merged_text=merged,
        has_conflicts=conflict_count > 0,
        conflict_count=conflict_count,
    )

File: ts-src/src/algo/merge.ts
import type { Edit } from './edit-index.js';
import { EditIndex } from './edit-index.js';

export interface MergeResult {
  readonly mergedText: string;
  readonly hasConflicts: boolean;
  readonly conflictCount: number;
}

function blocksToEdits(
  baseLen: number,
  otherLen: number,
  matchingBlocks: [number, number, number][],
): Edit[] {
  const edits: Edit[] = [];
  let prevI = 0;
  let prevJ = 0;
  for (const [i, j, n] of matchingBlocks) {
    if (prevI < i || prevJ < j) {
      edits.push([prevI, i, prevJ, j, false]);
    }
    if (n > 0) {
      edits.push([i, i + n, j, j + n, true]);
    }
    prevI = i + n;
    prevJ = j + n;
  }
  if (prevI < baseLen || prevJ < otherLen) {
    edits.push([prevI, baseLen, prevJ, otherLen, false]);
  }
  return edits;
}

/**
 * Simple longest-common-subsequence matching blocks (line-level).
 * Returns array of [baseIdx, otherIdx, length] tuples + terminal [baseLen, otherLen, 0].
 */
function getMatchingBlocks(base: string[], other: string[]): [number, number, number][] {
  const n = base.length;
  const m = other.length;

  // Build index: line → positions in other
  const otherIndex = new Map<string, number[]>();
  for (let j = 0; j < m; j++) {
    const line = other[j];
    if (line === undefined) throw new Error(`unreachable: other[${j}]`);
    let list = otherIndex.get(line);
    if (!list) {
      list = [];
      otherIndex.set(line, list);
    }
    list.push(j);
  }

  const blocks: [number, number, number][] = [];
  findBlocks({ base, otherIndex, baseLo: 0, baseHi: n, otherLo: 0, otherHi: m, result: blocks });
  blocks.sort((a, b) => a[0] - b[0] || a[1] - b[1]);
  blocks.push([n, m, 0]);
  return blocks;
}

interface FindBlocksParams {
  base: string[];
  otherIndex: Map<string, number[]>;
  baseLo: number;
  baseHi: number;
  otherLo: number;
  otherHi: number;
  result: [number, number, number][];
}

interface FindLongestBlockParams {
  base: string[];
  otherIndex: Map<string, number[]>;
  baseLo: number;
  baseHi: number;
  otherLo: number;
  otherHi: number;
}

function findLongestBlock(params: FindLongestBlockParams): {
  bestI: number;
  bestJ: number;
  bestLen: number;
} {
  const { base, otherIndex, baseLo, baseHi, otherLo, otherHi } = params;
  let bestI = baseLo;
  let bestJ = otherLo;
  let bestLen = 0;
  const runLens = new Map<number, number>();
  for (let i = baseLo; i < baseHi; i++) {
    const baseLine = base[i];
    if (baseLine === undefined) throw new Error(`unreachable: base[${i}]`);
    const newRuns = new Map<number, number>();
    for (const j of otherIndex.get(baseLine) ?? []) {
      if (j < otherLo || j >= otherHi) continue;
      const k = (runLens.get(j - 1) ?? 0) + 1;
      newRuns.set(j, k);
      if (k > bestLen) {
        bestI = i - k + 1;
        bestJ = j - k + 1;
        bestLen = k;
      }
    }
    runLens.clear();
    for (const [k, v] of newRuns) runLens.set(k, v);
  }
  return { bestI, bestJ, bestLen };
}

function findBlocks(params: FindBlocksParams): void {
  const { base, otherIndex, baseLo, baseHi, otherLo, otherHi, result } = params;
  const { bestI, bestJ, bestLen } = findLongestBlock({
    base,
    otherIndex,
    baseLo,
    baseHi,
    otherLo,
    otherHi,
  });

  if (bestLen === 0) return;

  if (baseLo < bestI && otherLo < bestJ) {
    findBlocks({ ...params, baseLo, baseHi: bestI, otherLo, otherHi: bestJ });
  }
  result.push([bestI, bestJ, bestLen]);
  if (bestI + bestLen < baseHi && bestJ + bestLen < otherHi) {
    findBlocks({
      ...params,
      baseLo: bestI + bestLen,
      baseHi,
      otherLo: bestJ + bestLen,
      otherHi,
    });
  }
}

function splitLines(text: string): string[] {
  if (!text) return [];
  return text.split(/\r\n|\r|\n/).map((l, i, a) => (i < a.length - 1 ? l + '\n' : l));
}

/**
 * Three-way merge for text content.
 *
 * Uses diff3 algorithm: computes edits from base→ours and base→theirs,
 * auto-merges non-overlapping changes, marks conflicts with markers.
 */
export function threeWayMerge(base: string, ours: string, theirs: string): MergeResult {
  const baseLines = splitLines(base);
  const oursLines = splitLines(ours);
  const theirsLines = splitLines(theirs);

  const mbOurs = getMatchingBlocks(baseLines, oursLines);
  const mbTheirs = getMatchingBlocks(baseLines, theirsLines);

  const editsOurs = blocksToEdits(baseLines.length, oursLines.length, mbOurs);
  const editsTheirs = blocksToEdits(baseLines.length, theirsLines.length, mbTheirs);

  const idxOurs = new EditIndex(editsOurs);
  const idxTheirs = new EditIndex(editsTheirs);

  const breakPoints = new Set<number>([0, baseLines.length]);
  for (const e of editsOurs) {
    breakPoints.add(e[0]);
    breakPoints.add(e[1]);
  }
  for (const e of editsTheirs) {
    breakPoints.add(e[0]);
    breakPoints.add(e[1]);
  }
  const pts = [...breakPoints].sort((a, b) => a - b);

  const segments = buildSegments(pts);
  const mergeCtx = {
    baseLines,
    oursLines,
    theirsLines,
    idxOurs,
    idxTheirs,
  };
  const { output, conflictCount } = mergeSegments(segments, mergeCtx);

  return {
    mergedText: output.join(''),
    hasConflicts: conflictCount > 0,
    conflictCount,
  };
}

function buildSegments(pts: number[]): [number, number][] {
  const segments: [number, number][] = [];
  for (let i = 0; i < pts.length; i++) {
    const p0 = pts[i];
    if (p0 === undefined) throw new Error(`unreachable: pts[${i}]`);
    if (i + 1 < pts.length) {
      const p1 = pts[i + 1];
      if (p1 === undefined) throw new Error(`unreachable: pts[${i + 1}]`);
      segments.push([p0, p0]);
      segments.push([p0, p1]);
    } else {
      segments.push([p0, p0]);
    }
  }
  if (segments.length === 0) segments.push([0, 0]);
  return segments;
}

interface MergeSegmentContext {
  baseLines: string[];
  oursLines: string[];
  theirsLines: string[];
  idxOurs: EditIndex;
  idxTheirs: EditIndex;
}

function mergeSegments(
  segments: [number, number][],
  ctx: MergeSegmentContext,
): { output: string[]; conflictCount: number } {
  const output: string[] = [];
  let conflictCount = 0;

  for (const [lo, hi] of segments) {
    if (lo > hi) continue;
    const oursInfo = ctx.idxOurs.find(lo, hi);
    const theirsInfo = ctx.idxTheirs.find(lo, hi);

    if (!oursInfo || !theirsInfo) {
      output.push(...ctx.baseLines.slice(lo, hi));
      continue;
    }

    const oursContent = ctx.oursLines.slice(oursInfo.otherLo, oursInfo.otherHi);
    const theirsContent = ctx.theirsLines.slice(theirsInfo.otherLo, theirsInfo.otherHi);
    const merged = mergeSegmentContent({
      lo,
      hi,
      oursInfo,
      theirsInfo,
      oursContent,
      theirsContent,
      baseLines: ctx.baseLines,
    });
    if (merged.conflict) conflictCount++;
    output.push(...merged.lines);
  }

  return { output, conflictCount };
}

interface MergeSegmentContentParams {
  lo: number;
  hi: number;
  oursInfo: { otherLo: number; otherHi: number; isSame: boolean };
  theirsInfo: { otherLo: number; otherHi: number; isSame: boolean };
  oursContent: string[];
  theirsContent: string[];
  baseLines: string[];
}

function mergeSegmentContent(params: MergeSegmentContentParams): {
  lines: string[];
  conflict: boolean;
} {
  const { lo, hi, oursInfo, theirsInfo, oursContent, theirsContent, baseLines } = params;
  if (oursInfo.isSame && theirsInfo.isSame) {
    return { lines: baseLines.slice(lo, hi), conflict: false };
  }
  if (oursInfo.isSame && !theirsInfo.isSame) {
    return { lines: theirsContent, conflict: false };
  }
  if (!oursInfo.isSame && theirsInfo.isSame) {
    return { lines: oursContent, conflict: false };
  }
  if (oursContent.join('') === theirsContent.join('')) {
    return { lines: oursContent, conflict: false };
  }
  return {
    lines: ['<<<<<<< LOCAL\n', ...oursContent, '=======\n', ...theirsContent, '>>>>>>> CLOUD\n'],
    conflict: true,
  };
}
